### ISSUE TITLE
New feature proposals

### ISSUE THREAD

**Issue by jchodera** [openmm/openmm#1980]
The lab discussed a number of potential new feature requests for the next release of OpenMM. I wanted to briefly summarize a number of them here. We can move ones that merit further discussion to separate issues.

#### Osmostat

We have a [draft implementation](https://github.com/choderalab/saltswap) and [preprint](https://doi.org/10.1101/226001) from @gregoryross describing an NCMC-based osmostat. We could build this into `Simulation` to automatically model macroscopic counterion concentrations. Some efficiency improvements would likely be required to make it practical for standard use.

#### Constant-pH

Similarly, we have a [draft implementation](https://github.com/choderalab/protons) of an NCMC-based protonation state sampling scheme from @bas-rustenburg. We're still working on a manuscript, so it may still be early for building this into `Simulation`, but we could think about whether we should refactor `Simulation` to support this and other plug-in MC modules (or if this should simply be built into [openmmtools](https://github.com/choderalab/openmmtools) instead).

#### Allow `NonbondedForce.updateParametersInContext()` to update subset of parameters

We have some preliminary profiling data from @bas-rustenburg that suggests that calls to `NonbondedForce.updateParametersInContext()` are a significant (~30%) component of constant-pH NCMC move costs. Since we only change a few particle parameters each step, there is no need to upload the whole array to the GPU; it may save time to request only a subset of particle parameters be updated.

cc: @bas-rustenburg 

#### Specify linear combination of parameters for `NonbondedForce` via context parameters

It would be useful for many applications (alchemical free energy calculations, constant-pH, osmostat, etc.) to be able to specify per-particle `NonbondedForce` parameters via a linear combination of per-particle constants,
```
q_i = \sum_{j=1}^M lambda_j q_{ij}
sigma_i = \sum_{j=1}^M lambda_j sigma_{ij}
epsilon_i = \sum_{j=1}^M lambda_j epsilon_{ij}
```
This would allow us to use full PME electrostatics and LJ, but modulate charges and Lennard-Jones parameters on the fly. Typically, we'd only need to define these linear combinations for a small portion of the system (a few ligand substitutents, some residues, or some salt/water molecules), rather than the whole system.
A variant of this proposal appears in https://github.com/pandegroup/openmm/issues/1702

cc: @andrrizzi

#### Copying/serializing `Integrator`s should copy their Python subclasses

Right now, copying or serializing a subclass of `CustomIntegrator` will not preserve the Python subclass. Allowing the Python subclass to be preserved would be highly useful for many cases, such as the integrators in [`openmmtools.integrators`](http://openmmtools.readthedocs.io/en/latest/integrators.html). 
cc: @andrrizzi https://github.com/pandegroup/openmm/issues/1911

#### Integrator plugin example/skeleton

Writing an [Integrator plugin](http://docs.openmm.org/latest/developerguide/developer.html#writing-plugins) is not at all trivial. Additional documentation and a skeleton example would be extremely helpful.

cc: @pgrinaway

#### Optimization of `Force` invalidation

Currently, changing a `Context` parameter will invalidate *all* forces, even if the parameter only affects a single `Custom*Force` that is inexpensive to recompute (such as a harmonic force where an umbrella center is being varied). Instead, individual `Force`s could be marked as valid/invalid to avoid recomputing energies or forces for those that don't require it when querying energies or forces at many different parameter values. This could provide enormous speedups for alchemical free energy and umbrella sampling calculations without any changes in the codes that use OpenMM.

cc: @andrrizzi

#### Providing a documented way to bypass `unit` wrapping

Since the overhead of unit wrapping is significant, for codes that use the OpenMM API to retrieve large numbers of parameters, it would be useful to provide a route to bypass the wrapping of quantities in `Unit` either by use of an additional optional Python argument or calling an alternate version of the function. This used to be possible by accessing the Swig function directly, but since some classes have been replaced at the Swig wrapper level, this is not uniformly available. This overhead is evident even in simple operations, such as retrieving the `System` masses or list of constrained atoms for a large solvated system (which can take several seconds, and in many applications, must be done over and over again).

cc: @pgrinaway

#### Retrieve/set `CustomIntegrator` kernel (CUDA or OpenCL) code

It would be useful to provide a route to retrieving the `.cu` or `.cl` code for `CustomIntegrator` kernels, and to possibly provide a way for the user to provide their own custom kernels if their computation does not fit within the constraints of `CustomIntegrator`.

cc: @pgrinaway

#### Neighborlists for `CustomNonbondedForce` with interaction groups

Currently, use of interaction groups in `CustomNonbondedForce` requires all interactions between both groups be computed every timestep. In alchemical free energy calculations, this can be quite expensive when only a fraction of the atoms are close enough to interaction. If neighborlists were used when the group size becomes sufficiently large, it could significantly reduce simulation times.

cc: @andrrizzi 

#### Allow more than 32 force groups

Some applications (such as linear basis function forcefields for simulating very large numbers of molecules at the same time) could benefit from more than 32 force groups. A way to provide more than these would be useful.

cc: @lnaden

**Comment by peastman** [comment]
> Osmostat

Definitely open an issue to discuss this.

> Constant-pH

We already have several issues open on that: #172, #1183, #1278.  Let's continue the discussion on those rather than opening yet another one.

 > Allow NonbondedForce.updateParametersInContext() to update subset of parameters

We already have an issue on this: #1661.  See the discussion there.  What you really mean is "improve the performance of updateParametersInContext()".  You're assuming the cost is dominated by uploading the new parameters to the GPU, but that's not always the case.

> Specify linear combination of parameters for NonbondedForce via context parameters

Is this basically the same as #1702?

> Copying/serializing Integrators should copy their Python subclasses

This is closely related to #1911.

> Integrator plugin example/skeleton

Closely related to #1873.

> Optimization of Force invalidation

This sounds like a good idea, and should be fairly easy to do.  Let's open an issue for it.

> Providing a documented way to bypass unit wrapping

This is #1490.  We never clearly established that this was really the problem.

> Retrieve/set CustomIntegrator kernel (CUDA or OpenCL) code

#1762.

> Neighborlists for CustomNonbondedForce with interaction groups

#1765.

> Allow more than 32 force groups

Related to #1973, though it's not exactly the same.

**Comment by jchodera** [comment]
Thanks! Will open/refresh issues as directed above.

One more idea we discussed in person:

#### Support for computation graphs

In order to facilitate interoperability with machine learning, or even just simplify the implementation of complex forces or CVs, the ability to create a `Force` from a computation graph (such as a [TensorFlow `tf.Graph`](https://www.tensorflow.org/api_docs/python/tf/Graph) or pytorch compute graph.

It may be difficult to support all Ops, but supporting a basic subset may still be useful.

**Comment by peastman** [comment]
> Support for computation graphs

Let's open an issue on that.  I think it's going to be a long discussion!

**Comment by jchodera** [comment]
One more important feature request I forgot!

#### RMSDForce

Many applications could benefit from use of a force dependent on the least-squares RMSD between subsets of atoms as a restraint or collective variable. In general, one would want to align on one specified set of coordinates while another set of coordinates is used to compute the least-squares RMSD.

I didn't find an open issue about this, but I might have missed it.

cc: @lnaden

**Comment by peastman** [comment]
Already implemented!  I'm just waiting until we declare 7.2 final to send the PR.

As I've implemented it currently, the alignment and RMSD are done on the same set of atoms.  Every paper I was able to find on calculating the derivatives of RMSD required that.  If you can find a derivation that doesn't need that, please point me to it.

**Comment by jchodera** [comment]
I think it might be straightforward to derive the necessary modifications. Which publication did you base your implementation on?

Suppose we align molecules on set `A` and measure the RMSD on set `B`:
```
RMSD = [(1/N_B) \sum_{i \in B} (U(x_A,xref_A) x_i - xref_i)' (U(x_A,xref_A) x_i - xref_i)]^{1/2}
```
where `x` and `xref` have been moved to their barycenters over atom subset `A`, and `U(x_A,xref_A)` is the optimal rotation matrix computed with atom subset `A`.

The derivative with respect to atom `x_j` is 
```
dRMSD/dx_j = (1/RMSD) (1/N_B) \sum_{i \in B} (U(x_A,xref_A) x_i - xref_i)' d/dx_j (U(x_A,xref_A) x_i - xref_i)
= (1/RMSD) \sum_{i \in B}^N (U(x_A,xref_A) x_i - xref_i)' (U(x_A,xref_A) \delta_{ij} + dU/dx_j x_i)
= (1/RMSD) \sum_{i \in B}^N (U(x_A,xref_A) x_i - xref_i)' U(x_A,xref_A) \delta_{ij}
+ (1/RMSD) \sum_{i \in B}^N (U(x_A,xref_A) x_i - xref_i)' dU(x_A,xref_A)/dx_j x_i
```
The first term involving `U(x_A,xref_A) \delta_{ij}` is simple: the optimal rotation matrix is computed using the subset of atoms `A`, and the only term that is included is the `j` term if `j \in B`; the term is zero otherwise.

The second term involves `dU(x_A,xref_A)/dx_j`, which will be zero unless `j \in A`. If it is, we should be able to adopt the existing result for the derivative of the rotation matrix with respect to one of the atom positions.

**Comment by peastman** [comment]
http://www.ams.stonybrook.edu/~coutsias/papers/rmsd17.pdf

**Comment by jchodera** [comment]
Great! I know Vageli. Which expression were you using?

**Comment by peastman** [comment]
The gradient is given by eq. 16.

**Comment by pgrinaway** [comment]
>Integrator plugin example/skeleton
>>Closely related to #1873.

Is this feasible?

**Comment by peastman** [comment]
Of course!

**Comment by tristanic** [comment]
How difficult would it be to apply the RMSDForce logic in dihedral space? That could be really useful (perhaps more so than Cartesian RMSD restraints - e.g. for long, flexible helices or coiled coil domains).

**Comment by jchodera** [comment]
Do you just mean that you want to somehow restrain specific torsions to be near a reference torsion? You can do that right now with CustomCompoundBondForce or CustomTorsionForce:

* http://docs.openmm.org/latest/api-python/generated/simtk.openmm.openmm.CustomTorsionForce.html#simtk.openmm.openmm.CustomTorsionForce
* http://docs.openmm.org/latest/api-python/generated/simtk.openmm.openmm.CustomCompoundBondForce.html#simtk.openmm.openmm.CustomCompoundBondForce

Just make sure you use a potential that's periodic over the right domain!

**Comment by tristanic** [comment]
No, I can already do that - and perhaps that's all that's needed, if used carefully. 

... actually, come to think of it, that *is* all that's needed, isn't it? Never mind...

**Comment by tristanic** [comment]
A feature that I suspect would find reasonably wide use would be array forms of force parameter setters (e.g. setBondParameters()) to allow fast changes from Python. I'm currently implementing my own little extension library for this, but it wouldn't be a hugely arduous task to implement directly in the SWIG wrapper code using %extend and %pythoncode.

An example of what I'm talking about:

In C++:
```
extern "C" void
custombondforce_update_bond_parameters(void *force, size_t n, int *indices, double *params)
{
    OpenMM::CustomBondForce *f = static_cast<OpenMM::CustomBondForce *>(force);
    try {
        int n_params = f->getNumPerBondParameters();
        std::vector<double> param_vec(n_params);
        int particle1, particle2;
        for (size_t i=0; i<n; ++i) {
            int index = *(indices++);
            f->getBondParameters(index, particle1, particle2, param_vec);
            for (int j=0; j<n_params; ++j) {
                param_vec[j] = *params++;
            }
            f->setBondParameters(index, particle1, particle2, param_vec);
        }
    } catch (...) {
        molc_error();
    }
}
```
... where molc_error() is just a simple function to pass exceptions from C++ to Python.

Then in my Python subclass of CustomBondForce, I can do (using a few wrapper functions to ctypes, which wouldn't be necessary using SWIG):

```
    def update_targets(self, indices, targets, spring_constants):
        '''
        Update a set of targets all at once using fast C++ code. Fastest if
        the arguments are provided as Numpy arrays, but any iterable will work.
        @param indices:
            The indices of the restraints in the OpenMM force object
        @param targets:
            The new target distances in nanometres
        @param k
            The new spring constants in units of kJ mol-1 nm-2
        '''
        f = c_function('custombondforce_update_bond_parameters',
            args=(ctypes.c_void_p, ctypes.c_size_t,
            ctypes.POINTER(ctypes.c_int32), ctypes.POINTER(ctypes.c_double)))
        n = len(indices)
        if len(targets) != n or len(spring_constants) != n:
            raise TypeError('Parameter array lengths must match number of indices!')
        ind = numpy.empty(n, int32)
        ind[:] = indices
        params = numpy.empty((n,2), float64)
        params[:,0] = targets
        params[:,1] = spring_constants
        f(int(self.this), n, pointer(ind), pointer(params))
        self.update_needed = True
```

... which, when all put together, allows me to do, for example:

```
from chimerax.isolde.openmm import custom_forces
tbf = custom_forces.TopOutBondForce(2000)
force_indices = []
indices = [[i,j] for i, j in zip(range(10), range(10,20))]
for pair in indices:
    force_indices.append(tbf.addBond(*pair, (1.0, 1.0)))
for i in force_indices:
    print(tbf.getBondParameters(i))
import numpy
targets = numpy.random.rand(10)
ks = numpy.random.rand(10)
tbf.update_targets(force_indices, targets, ks)
for i in force_indices:
    print(tbf.getBondParameters(i))

[0, 10, (1.0, 1.0)]
[1, 11, (1.0, 1.0)]
[2, 12, (1.0, 1.0)]
[3, 13, (1.0, 1.0)]
[4, 14, (1.0, 1.0)]
[5, 15, (1.0, 1.0)]
[6, 16, (1.0, 1.0)]
[7, 17, (1.0, 1.0)]
[8, 18, (1.0, 1.0)]
[9, 19, (1.0, 1.0)]
[0, 10, (0.9892172794881968, 0.90526046406322)]
[1, 11, (0.970934906811111, 0.6236264019114641)]
[2, 12, (0.9670802612656184, 0.6142107827470832)]
[3, 13, (0.38402119109091026, 0.5437964523374115)]
[4, 14, (0.3907161962307253, 0.9794193836195636)]
[5, 15, (0.8685141649243204, 0.574622250992786)]
[6, 16, (0.2506946551065128, 0.7710411756497542)]
[7, 17, (0.12761334365957944, 0.6918066369703049)]
[8, 18, (0.8030626730631322, 0.4488705539824319)]
[9, 19, (0.9755068820070552, 0.9287106019559775)]
```

For 10,000 bonds, setting all the bond parameters in Python using the existing calls takes 43 ms, whereas the above code takes 1.3 ms. Whether that's an important speedup of course depends entirely on the application - but it also substantially simplifies the Python code, and sidesteps some really annoying interactions between SWIG and Numpy scalars (i.e. it refuses to recognise numpy.int32 scalars until they've been explicitly cast to Python ints).

**Comment by tristanic** [comment]
The ability to update/change a TabulatedFunction in a CustomCompoundBondForce would also be really useful. My use case for this is in MDFF-based crystallographic refinement, where the electron density map isn't a constant but needs to be periodically recalculated based on the current coordinates. Being able to update the map on the fly would be wonderful.
